server/http/ipfs.go

```diff
@@ -105,67 +106,78 @@ func ipfsHandler(lassie *lassie.Lassie) func(http.ResponseWriter, *http.Request)
 		// }
 		// TODO: Do something with unixfs path
 
-		log.Debug("creating temp car file")
-		carfile, err := os.CreateTemp("", rootCid.String())
-		if err != nil {
-			logger.LogStatus(http.StatusInternalServerError, fmt.Sprintf("Failed to create temp car file before retrieval: %v", err))
-			res.WriteHeader(http.StatusInternalServerError)
-			return
+		// for setting Content-Disposition header based on filename url parameter
+		var filename string
+		if req.URL.Query().Has("filename") {
+			filename = req.URL.Query().Get("filename")
+		} else {
+			filename = fmt.Sprintf("%s.car", rootCid.String())
 		}
 
-		// create blockstore from defined car file and use it for the link system
-		blockstore, err := blockstore.OpenReadWriteFile(carfile, []cid.Cid{rootCid}, blockstore.WriteAsCarV1(true))
+		retrievalId, err := types.NewRetrievalID()
 		if err != nil {
-			logger.LogStatus(http.StatusInternalServerError, fmt.Sprintf("Failed to create blockstore from temp car file before retrieval: %v", err))
-			res.WriteHeader(http.StatusInternalServerError)
+			msg := fmt.Sprintf("Failed to generate retrieval ID: %s", err.Error())
+			logger.LogStatus(http.StatusInternalServerError, msg)
+			http.Error(res, msg, http.StatusInternalServerError)
 			return
 		}
-		linkSystem := storeutil.LinkSystemForBlockstore(blockstore)
 
-		log.Debugw("fetching cid into car file", "cid", rootCid.String(), "file", carfile.Name())
-		id, _, err := lassie.Fetch(req.Context(), rootCid, linkSystem)
-		if err != nil {
-			logger.LogStatus(http.StatusInternalServerError, fmt.Sprintf("Failed to fetch cid: %v", err))
-			res.WriteHeader(http.StatusInternalServerError)
-			carfile.Close()
-			return
+		// called once we start writing blocks into the CAR (on the first Put())
+		getWriter := func() (io.Writer, error) {
```

Comment:
ok I was reading the previous code being like "why does he have a getWriter method" and now it makes a LOT of sense. Man I dunno if there's a better way to do this but this is at least a big improvement over listening directly for the first byte event which seemed awkward. I'm not sure what the ultimate best method is, but this seems good enough for now.